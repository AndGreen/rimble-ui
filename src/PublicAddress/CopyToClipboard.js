import React, { useState, useEffect, useRef } from 'react';
import ClipboardJS from 'clipboard';

const Clipboard = props => {
  const [isCopied, setIsCopied] = useState(false);
  const targetEl = useRef(null);

  useEffect(() => {
    let timer;

    const clipboard = new ClipboardJS(targetEl.current, {
      // target: () => (),
      text: () => props.text,
    });

    clipboard.on('success', e => {
      // console.log('text copied!:', props.text);
      setIsCopied(true);
      clearTimeout(timer);
      timer = setTimeout(() => setIsCopied(false), 1500);
    });

    clipboard.on('error', e => {
      console.log('error: failed to copy text');
    });

    return () => {
      clipboard.destroy();
      clearTimeout(timer);
    };
  });

  return <div ref={targetEl}>{props.children(isCopied)}</div>;
};

export default Clipboard;
