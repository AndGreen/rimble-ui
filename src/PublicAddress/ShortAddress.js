import React, { Component } from 'react';

const addressSummary = (address, firstSegLength = 10, lastSegLength = 4, includeHex = true) => {
  if (!address) return ''
  // let checked = checksumAddress(address)
  // if (!includeHex) {
  //   checked = ethUtil.stripHexPrefix(checked)
  // }
  return address ? address.slice(0, firstSegLength) + '...' + address.slice(address.length - lastSegLength) : '...'
}

class ShortAddress extends Component {

  render() {
    return (
      <div>{addressSummary(this.props.address)}</div>
    );
  }

}

export default ShortAddress;
