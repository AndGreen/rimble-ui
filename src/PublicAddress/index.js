import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Box from '../Box';
import Flex from '../Flex';
import Card from '../Card';
import Text from '../Text';
import Tooltip from '../Tooltip';

import Icon from '../Icon';
import Button from '../Button';
import Input from '../Input';
import EthAddress from '../EthAddress';
import QR from '../QR';
import Portal from '../Portal';
import { useHiddenState } from '../Hidden';
import { ModalBackdrop } from '../Modal';

const StyledInput = styled(Input)`
  text-overflow: ellipsis;
  white-space: no-wrap;
`;

const StyledWrapper = styled(Box)`
  & {
    display: flex;
    flex-direction: row;
    align-items: center;
    width: 100%;
    position: relative;
  }

  ${'' /* > input {
    width: 100%;
    padding-right: 5.25rem;
  } */}

  ${'' /* > button {
    position: absolute;
    width: 4rem;
    right: 0.75rem;
  } */}
`;

const AddressQrModal = ({ isOpen, hide, address }) => {
  if (isOpen) {
    return (
      <Portal>
        <ModalBackdrop>
          <Card
            width={'auto'}
            m={3}
            border={'none'}
            borderRadius={2}
            bg={'primary-light'}
          >
            <Text color={'white'} mb={4} textAlign={'center'}>
              Use your camera to scan the code.
            </Text>

            <Box
              bg={'white'}
              size={'200px'}
              mx={'auto'}
              mb={4}
              p={3}
              boxShadow={2}
            >
              <QR value={address} size={'100%'} />
            </Box>

            <Text
              color={'white'}
              bg={'primary'}
              px={4}
              py={3}
              borderRadius={2}
              fontWeight={3}
              lineHeight={'solid'}
            >
              {address}
            </Text>

            <Box position={'absolute'} top={0} right={0}>
              <Button.Text
                icon={'Close'}
                mainColor={'white'}
                m={1}
                p={0}
                borderRadius={'100%'}
                onClick={hide}
              />
            </Box>
          </Card>
        </ModalBackdrop>
      </Portal>
    );
  }

  return null;
};

const QRButton = ({ address }) => {
  const { visible, toggle } = useHiddenState();
  return (
    <React.Fragment>
      <Tooltip message={'show QR code'}>
        <Button size={'small'} ml={2} p={0} onClick={toggle}>
          <Icon name={'CenterFocusStrong'} />
        </Button>
      </Tooltip>
      <AddressQrModal address={address} isOpen={visible} hide={toggle} />
    </React.Fragment>
  );
};

class PublicAddress extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isCopied: false,
    };
    this.inputRef = React.createRef();
    this.buttonRef = React.createRef();
  }

  handleClick = e => {
    const { inputRef, buttonRef } = this;

    e.preventDefault();
    inputRef.current.select();
    document.execCommand('copy');
    this.setState({ isCopied: true });

    setTimeout(() => {
      this.setState({ isCopied: false });
    }, 5000);
  };

  render() {
    return (
      <StyledWrapper>
        <StyledInput
          readOnly
          value={this.props.address}
          ref={this.inputRef}
          width={1}
          pr={'6rem'}
        />
        {/* <EthAddress address={this.props.address} /> */}

        <Flex position={'absolute'} right={0} mr={2}>
          <Tooltip message={'copy to clipboard'}>
            <Button
              size={'small'}
              ml={2}
              p={0}
              onClick={this.handleClick}
              ref={this.buttonRef}
            >
              <Icon name={this.state.isCopied ? 'Check' : 'Assignment'} />
            </Button>
          </Tooltip>
          <QRButton address={this.props.address} />
        </Flex>
      </StyledWrapper>
    );
  }
}

PublicAddress.propTypes = {
  /**
   * Sets Ethereum address as the value of the field
   */
  address: PropTypes.string,
};

PublicAddress.displayName = 'PublicAddress';

export default PublicAddress;
