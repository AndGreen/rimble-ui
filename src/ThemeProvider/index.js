import React from 'react'
import PropTypes from 'prop-types'

import styled, {
  ThemeProvider as StyledThemeProvider
} from 'styled-components'

import theme from '../theme'

export const Base = styled.div`
  font-family: ${props => props.theme.fonts.sansSerif};
  font-size: 16px;
  * {
    box-sizing: border-box;
  }
`

const ThemeProvider = ({ ...props }) => {
  return (
    <StyledThemeProvider theme={theme}>
      <Base {...props} />
    </StyledThemeProvider>
  )
}

ThemeProvider.propTypes = {
}

export default ThemeProvider
