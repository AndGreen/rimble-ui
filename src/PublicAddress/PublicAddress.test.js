import PublicAddress from './'
import ThemeProvider from '../ThemeProvider'
import theme from '../theme'

describe('PublicAddress component sanity', () => {
  it('has name', () => {
    expect(PublicAddress.displayName).toBe('PublicAddress')
  })

  it('matches themed snapshot', () => {
    const component = render(
      <ThemeProvider theme={theme}>
        <PublicAddress />
      </ThemeProvider>
    )
    expect(component).toMatchSnapshot()
  })
})
