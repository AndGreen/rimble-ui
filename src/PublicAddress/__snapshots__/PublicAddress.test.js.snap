// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublicAddress component sanity matches default snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0 {
  box-sizing: border-box;
}

.c5 {
  margin: 0px;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 0px;
  padding-bottom: 0px;
  color: #fff;
  background-color: #4E3FCE;
  font-size: inherit;
  height: 48px;
  height: small;
  width: small;
  position: relative;
  border: 0;
  border-color: none;
  border-radius: 4px;
}

.c5 {
  box-sizing: border-box;
}

.c1 {
  margin-bottom: 8px;
  color: copyColor;
  font-size: 14px;
  font-size: 14px;
  font-family: sansSerif;
  font-weight: 3;
  text-align: left;
  line-height: copy;
}

.c1 {
  box-sizing: border-box;
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-size: 1rem;
  line-height: 1;
  height: 48px;
  padding: 0 1rem;
  border: 1px solid;
  background: #FFFFFF;
  border-color: #CCC;
  border-radius: 4px;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
  text-overflow: ellipsis;
  white-space: no-wrap;
}

.c3:hover {
  box-shadow: 0px 2px 6px rgba(0,0,0,0.3);
}

.c3:focus {
  outline: none;
  border-color: #5436D6;
}

.c3:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c4 {
  font-weight: 600;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
  font-size: 12px;
  height: 32px;
  min-width: 32px;
  padding: 0 16px;
}

.c4 {
  cursor: pointer;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  line-height: 1;
  overflow: hidden;
  white-space: nowrap;
  -webkit-font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all .15s ease;
  transition: all .15s ease;
}

.c4:hover {
  background-color: #5f52d2;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.2);
}

.c4:active {
  background-color: #4638b9;
  box-shadow: none;
}

.c4:disabled {
  opacity: 0.4;
  cursor: not-allowed;
}

.c4 > svg {
  display: block;
}

.c4 > svg:first-child {
  margin-left: -1rem;
  margin-right: .5rem;
}

.c4 > svg:last-child {
  margin-left: .5rem;
  margin-right: -1rem;
}

.c4 > svg:first-child:last-child {
  margin: 0;
}

.c4 {
  position: relative;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4.-is-copied::before {
  content: '';
  position: absolute;
  width: 100%;
  height: 100%;
  background: inherit;
  opacity: 1;
  -webkit-animation-name: eMLfYp;
  animation-name: eMLfYp;
  -webkit-animation-duration: 100ms;
  animation-duration: 100ms;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
}

.c4::after {
  opacity: 0;
  content: 'Copied!';
  position: absolute;
  width: 100%;
  background: inherit;
  -webkit-transition: all 500ms ease;
  transition: all 500ms ease;
}

.c4.-is-copied::after {
  opacity: 1;
  -webkit-animation-name: kcJLrX;
  animation-name: kcJLrX;
  -webkit-animation-duration: 1000ms;
  animation-duration: 1000ms;
  -webkit-animation-timing-function: ease;
  animation-timing-function: ease;
}

.c2 {
  box-sizing: border-box;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  position: relative;
}

.c2 > input {
  width: 100%;
  padding-right: 5.25rem;
}

.c2 > button {
  position: absolute;
  right: .75rem;
}

<label
  class="c0"
  display="flex"
>
  <div
    class="c1"
    color="copyColor"
    font-family="sansSerif"
    font-size="1"
    font-weight="3"
  >
    Public Address
  </div>
  <div
    class="c2"
  >
    <input
      class="c3"
      readonly=""
    >
      
    </input>
    <button
      class="c4 c5"
      color="white"
      font-size="inherit"
      font-weight="3"
      height="48px"
      type="button"
    >
      <span>
        Copy
      </span>
    </button>
  </div>
</label>
`;
